### PR TITLE
feat(cli) kong backup CLI command

### LINKED ISSUE #1374: API endpoint to export and import all the DB entities
One of the most common requests is to move data from one cluster to another (or one environment to another) in an easy way. Currently the two ways this can be achieved are:
- At the datastore level, by migrating the keyspace or the database
- By replicating the same HTTP requests to the Admin API on both clusters

Kong could possibly facilitate this by providing `/export` and `/import` endpoints which basically just dump the entire content of the datastore in a JSON file, ready to be imported by another cluster.

### PR DESCRIPTION
**Pull request by thefosk**
### Summary

Implements a `kong backup [create/import]` CLI command to create and import backups of the data in Kong, regardless of the datastore being used.

This command can also be used to move data around different Kong clusters.
### Full changelog
- New `kong backup [create/import]` CLI command to create and import backups of data stored in a Kong cluster.
### Issues resolved

Closes #1374.

### CODE REVIEW

File: spec/02-integration/01-cmd/11-backup_spec.lua [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        local ok, err = helpers.kong_exec("backup import -y --conf "..helpers.test_conf_path)
+        assert.False(ok)
+        assert.matches("Error: must specify the folder path to import", err, nil, true)
+      end)
+      it("fails with an invalid folder", function()
+        local ok, err = helpers.kong_exec("backup import /tmp -y --conf "..helpers.test_conf_path)
+        assert.False(ok)
+        assert.matches("Error: Backup is missing the metadata file", err, nil, true)
+      end)
+    end)
```
Comment (Tieske) [inline spec/02-integration/01-cmd/11-backup_spec.lua]: test; it should fail with a different kong version.

File: kong/cmd/backup.lua [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Create or import backups of data stored in Kong.
+
+The available commands are:
+ create            Create a new backup from the database
+ import <folder>   Import an existing backup into the database
+
+Options:
+ -c,--conf (optional string) configuration file
+ -y        Assume yes; assume that the answer to any question which would be asked is yes
+]]
```
Comment (Tieske) [inline kong/cmd/backup.lua]: I think it requires a password option to encrypt the output

**Comment by Tieske** [comment]
also; what if importing, but the database is not empty? should that be allowed?

**Comment by thefosk** [comment]
> what if importing, but the database is not empty? should that be allowed?

Maybe an error message can be returned if this is the case.

> encrypting the output

out of scope. The system generates a backup, up to the user to decide what to do it with it and optionally encrypt it.

> shouldn't it be using the mgt api?

No, this is much more performant since it accesses the DAO directly.

**Comment by oviwankenobi** [comment]
hi, thanks for all your work, any idea on the progress of it, when this feature will be ready?

**Comment by Tieske** [comment]
Just ran some tests;

- importing aborts when it hits an existing uuid, leaving everything in an undetermined state
- previous point raises the question; should an import on a non-empty db be allowed
- resetting the datastore does not work. Because the import doesn't run the migrations
- previous point should be addressed by storing the db schema version used to create the backup and restore everything only on an empty version of that db schema.
- still questioning the necessity of a `validate` option to verify everything was exported ok (to make sure no intermediate changes rendered the backup inconsistent)
- should an import be allowed on a running Kong node?

**Comment by kkindaface** [comment]
I tried a 'kong backup' on 0.11.0 release and it did not show up in the available options from command line, has this feature been removed?

**Comment by hishamhm** [comment]
> I tried a 'kong backup' on 0.11.0 release and it did not show up in the available options from command line, has this feature been removed?

No, it has never been added.

**Comment by thefosk** [comment]
@kkindaface this feature is being rebuilt properly in the EE edition, with future support for recurring cloud backups as well. If you are interested please [contact us](https://www.mashape.com/enterprise/).

**Comment by thibaultcha** [comment]
@thefosk Can we remove this branch from this repository? It can be pushed elsewhere if needed. The repo is currently cluttered with a bunch stale branches.

**Comment by kkindaface** [comment]
Not a problem, I tried using it after I saw the command mentioned in the docs page with no description. https://getkong.org/docs/0.11.x/cli/#kong-backup

Thanks for the response!

**Comment by thefosk** [comment]
@kkindaface thanks for the heads up, somehow it slipped in there

**Pull request closed without merging**